src/core/lib/event_engine/posix_engine/posix_engine.cc

```diff
@@ -46,7 +58,221 @@ namespace grpc_event_engine {
 namespace experimental {
 
 #ifdef GRPC_POSIX_SOCKET_TCP
+using grpc_event_engine::posix_engine::EventHandle;
+using grpc_event_engine::posix_engine::PosixEngineClosure;
 using grpc_event_engine::posix_engine::PosixEventPoller;
+using grpc_event_engine::posix_engine::PosixSocketWrapper;
+using grpc_event_engine::posix_engine::PosixTcpOptions;
+using grpc_event_engine::posix_engine::SockaddrToString;
+using grpc_event_engine::posix_engine::TcpOptionsFromEndpointConfig;
+
+void AsyncConnect::Start(EventEngine::Duration timeout) {
+  on_writable_ = PosixEngineClosure::ToPermanentClosure(
+      [this](absl::Status status) { OnWritable(std::move(status)); });
+  alarm_handle_ = engine_->RunAfter(timeout, [this]() {
+    OnTimeoutExpired(absl::DeadlineExceededError("connect() timed out"));
+  });
+  fd_->NotifyOnWrite(on_writable_);
+}
+
+AsyncConnect ::~AsyncConnect() { delete on_writable_; }
+
+void AsyncConnect::OnTimeoutExpired(absl::Status status) {
+  bool done = false;
+  {
+    absl::MutexLock lock(&mu_);
+    if (fd_ != nullptr) {
+      fd_->ShutdownHandle(std::move(status));
+    }
+    done = (--refs_ == 0);
+  }
+  if (done) {
+    delete this;
+  }
+}
+
+void AsyncConnect::OnWritable(absl::Status status)
+    ABSL_NO_THREAD_SAFETY_ANALYSIS {
+  int so_error = 0;
+  socklen_t so_error_size;
+  int err;
+  int done;
+  int consumed_refs = 1;
+  EventHandle* fd;
+  absl::StatusOr<std::unique_ptr<EventEngine::Endpoint>> ep;
+
+  mu_.Lock();
+  GPR_ASSERT(fd_ != nullptr);
+  fd = std::exchange(fd_, nullptr);
+  bool connect_cancelled = connect_cancelled_;
+  mu_.Unlock();
+
+  if (engine_->Cancel(alarm_handle_)) {
```

Comment:
Is there a race here between OnWritable and OnTimeoutExpired? If OnWritable is started in one thread, but OnTimeoutExpires enters its critical section first, then I think AsyncConnect will be deleted while OnWritable is running.

Reply:
I dont think so. If engine_->Cancel succeeds, then the OnTimeoutExpires callback shouldn't run. So it wont decrement the refcount. Instead we just increment the consumed_refs variable within the if block. If OnTimeoutExpires is running while OnWritable is also executing, the engine_->Cancel should have returned false so there shouldn't be an extra increment of the ref count. The rest of  OnWritable would also be executing under the locked mutex mu_

Reply:
I see, there are 2 `refs_` to start, so it won't be deleted in the cancel callback.